### PR TITLE
Field visibility 

### PR DESCRIPTION
**Pull request by bbakerman**
This adds the ability to filter the fields that are visible from a graphql schema at runtime.

The schema is built with the complete 100% set of fields and then field filtering can be applied to make them disappear

So if you make "secretField" filtered out the following will be true

Validation will fail with an unknown field if they ask for it
Execution wont ever load the field as a value if it makes it pase validation
Introspection won't see the field and hence report it as being a field you can query

It provides 2 pre-built visibility a simple black list on name implementation and black list introspection at all implementation.

### CODE REVIEW

File: src/main/java/graphql/schema/GraphqlFieldVisibility.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    GraphQLFieldDefinition getFieldDefinition(GraphQLFieldsContainer fieldsContainer, String fieldName);
+
+    /**
+     * Called to get the list of enum values of an enum type
+     *
+     * @param enumType the enum type in play
+     *
+     * @return a non null list of {@link graphql.schema.GraphQLEnumValueDefinition}s
+     */
+    List<GraphQLEnumValueDefinition> getValues(GraphQLEnumType enumType);
```
Comment (andimarek) [inline src/main/java/graphql/schema/GraphqlFieldVisibility.java]: do we need to restrict that at all? Isn't field restriction not enough?

Comment (bbakerman) [inline src/main/java/graphql/schema/GraphqlFieldVisibility.java]: removed

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 2]
```diff
@@ -586,7 +586,7 @@ protected GraphQLFieldDefinition getFieldDef(GraphQLSchema schema, GraphQLObject
             return TypeNameMetaFieldDef;
         }
 
-        GraphQLFieldDefinition fieldDefinition = parentType.getFieldDefinition(field.getName());
+        GraphQLFieldDefinition fieldDefinition = schema.getFieldVisibility().getFieldDefinition(parentType, field.getName());
         if (fieldDefinition == null) {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionStrategy.java]: This should not be possible if validation catches the unknown field (which we have tests for) but belts and braces

File: src/main/java/graphql/schema/visibility/BlacklistGraphqlFieldVisibility.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * This helper class will take a list of regular expressions and match them against the fully qualified name
+ * of a type and its fields.  So for example an object type called "User" with an inner field called "firstName"
+ * will have a fully qualified name of "User.firstName" in terms of pattern matching.
+ *
+ * Remember that graphql type and fields names MUST be inside the name space "[_A-Za-z][_0-9A-Za-z]*"
+ */
+@PublicApi
+public class BlacklistGraphqlFieldVisibility implements GraphqlFieldVisibility {
```
Comment (andimarek) [inline src/main/java/graphql/schema/visibility/BlacklistGraphqlFieldVisibility.java]: Could we rename it to `BlockedGraphqlFieldVisibility`?

Comment (andimarek) [inline src/main/java/graphql/schema/visibility/BlacklistGraphqlFieldVisibility.java]: Or maybe just `BlockedFields` ...

Comment (bbakerman) [inline src/main/java/graphql/schema/visibility/BlacklistGraphqlFieldVisibility.java]: np

**Comment by bbakerman** [comment]
I have just realised this is missing a key bit of context.  There is no way to get the "context" object
on the call to filter a field and hence there is no way to make a smart decision based on anything else but name

We need to pass the "context" object into them so they can access their user info say or whatever.

**Comment by kaqqao** [comment]
Just to double-check if I get it right. The visibility isn't meant to be dynamic (as in: the same schema with the same visibility strategy used for everyone and dynamically deciding who gets to see what on the fly), right? You're supposed to create a schema per group/role/user with different visibilities instead (and it's a cheap operation as all the same references are reused)?
Going further, I reckon it is done to ensure schema validity, as if it was dynamically deciding (vs the schema build time), it could easily produce an invalid schema?

**Comment by bbakerman** [comment]
> @bbakerman Just to double-check if I get it right. The visibility isn't meant to be dynamic (as in: the same schema with the same visibility strategy used for everyone and dynamically deciding who gets to see what on the fly), right? You're supposed to create a schema per group/role/user with different visibilities instead (and it's a cheap operation as all the same references are reused)?
Going further, I reckon it is done to ensure schema validity, as if it was dynamically deciding (vs the schema build time), it could easily produce an invalid schema?

Actually no - you can do it either way.  You can build a schema per visibility object.  Or you can have it dynamic.

If you make your schema invalid because of whacky field exclusion, be it on your head.

This will dynamically filter fields when asked.  So it can be as dynamic or static as you choose.

**Comment by emrul** [comment]
I'd been off our project for a while but am now back on it and I have to say, on initial glance this work looks really, really great.  I think we can get behind this and kick-away our custom fork - thank you @bbakerman & @andimarek !

**Comment by breathermachine** [comment]
Hello, I'm new to GraphQL and I'm trying to use this functionality to perform per-field authorization. Is there a way to customize the error message?

Instead of "Validation error of type FieldUndefined: Field 'name' in type 'Organization' is undefined @ 'employees/homePosition/organization/name'", I would like to error to be something like:
"Validation error of type FieldAccessError: Field 'name' in type 'Organization' is not accessible given current credentials"

Thank you!

**Comment by kaqqao** [comment]
@breathermachine As far as I know, you can not use `GraphQLFieldVisibility` like that. But I reckon you could achieve something similar using an instrumentation with customized `beginField` (that would perhaps throw an `AbortExecutionException`).
That said, it's a general security guideline (neither Java nor GraphQL specific) _not_ to give different error messages for _not found_ and _found but not accessible_ as that gives the potential attacker better insight into your system. Still, this is not always applicable, so make whatever your system requires.

**Comment by breathermachine** [comment]
@kaqqao Thank you for your response. Actually what I wanted to do was to return a 403 while revealing as little information to the outside as possible. However, graphql is returning the same error message for fields not supported by the schema, and blocked fields. I would like to have it distinguish the two cases so I can send the appropriate HTTP code (404 for the first case and 403 for the second).

**Pull request merged**